### PR TITLE
Proper deregistration of static types and objects

### LINKED ISSUE #61111: [Windows] Crash when unloading torch_cpu.dll using LOCAL SYSTEM account
## 🐛 Bug

Somewhat similarly to https://github.com/pytorch/pytorch/issues/19055, we have an application plugin which links against libtorch. The plugin is loaded and unloaded on demand. The application crashes (`Access violation reading location ...`) when the plugin is unloaded. The issue can be observed when running the application using LOCAL SYSTEM account (e.g. as a system service).

## To Reproduce

Consider the following MWE:

### torch_client.cc
```
#include <torch/script.h>

void fun() { torch::Device device("cpu"); } 
```

### demo.cc
```
#include <boost/dll.hpp>
#include <iostream>

static const auto kPluginLoadMode =
    boost::dll::load_mode::append_decorations |
    boost::dll::load_mode::search_system_folders;

void test() {
    auto torch_client =
        boost::dll::shared_library("TorchClient", kPluginLoadMode);
    char c;
    std::cin >> c;
}

int main() try {
    test();
    return 0;
} catch (std::exception& e) {
    std::cerr << "Fatal: " << e.what() << "\n";
    return 1;
} 
```

### CMakeLists.txt
```
...
find_package(Torch REQUIRED)
find_package(Boost REQUIRED COMPONENTS system filesystem)

add_library(TorchClient SHARED torch_client.cc)
target_link_libraries(TorchClient PRIVATE torch)

add_executable(Demo demo.cc)
target_link_libraries(Demo
    PRIVATE Boost::system
    PRIVATE Boost::filesystem) 
```

1. Put TorchClient.dll, Demo.exe and all of PyTorch DLLs and its direct and indirect dependencies in some directory.
1. Open LOCAL SYSTEM shell by running `psexec -i -s cmd` (from SysInternals suite).
1. Run Demo.exe from there.

This results in an application error event logged
```
Faulting application name: Demo.exe, version: 0.0.0.0, time stamp: 0x60dd819d
Faulting module name: torch_cpu.dll, version: 0.0.0.0, time stamp: 0x60db1358
Exception code: 0xc0000005
Fault offset: 0x0000000005815e42
Faulting process id: 0x140
Faulting application start time: 0x01d76e59eebd0eb4
Faulting application path: C:\Users\dorozhkin\Projects\qa_tests\test_xx\Demo.exe
Faulting module path: C:\Users\dorozhkin\Projects\qa_tests\test_xx\torch_cpu.dll
Report Id: f9593fc6-d26b-42b4-9e67-2f175bf752e1
Faulting package full name: 
Faulting package-relative application ID: 
```

Stacktrace collected from running `devenv /debugexe`:
```
>	[Inline Frame] torch_cpu.dll!std::_Func_class<bool,torch::jit::Node const *>::_Tidy() Line 1046	C++
 	[Inline Frame] torch_cpu.dll!std::_Func_class<bool,torch::jit::Node const *>::{dtor}() Line 972	C++
 	[Inline Frame] torch_cpu.dll!std::_Default_allocator_traits<std::allocator<std::function<bool __cdecl(torch::jit::Node const *)>>>::destroy(std::allocator<std::function<bool __cdecl(torch::jit::Node const *)>> &) Line 723	C++
 	[Inline Frame] torch_cpu.dll!std::_Destroy_range(std::function<bool __cdecl(torch::jit::Node const *)> * _First, std::function<bool __cdecl(torch::jit::Node const *)> * const) Line 950	C++
 	[Inline Frame] torch_cpu.dll!std::vector<std::function<bool __cdecl(torch::jit::Node const *)>,std::allocator<std::function<bool __cdecl(torch::jit::Node const *)>>>::_Destroy(std::function<bool __cdecl(torch::jit::Node const *)> * _Last, std::function<bool __cdecl(torch::jit::Node const *)> *) Line 1680	C++
 	[Inline Frame] torch_cpu.dll!std::vector<std::function<bool __cdecl(torch::jit::Node const *)>,std::allocator<std::function<bool __cdecl(torch::jit::Node const *)>>>::_Tidy() Line 1764	C++
 	torch_cpu.dll!std::vector<std::function<bool __cdecl(torch::jit::Node const *)>,std::allocator<std::function<bool __cdecl(torch::jit::Node const *)>>>::~vector<std::function<bool __cdecl(torch::jit::Node const *)>,std::allocator<std::function<bool __cdecl(torch::jit::Node const *)>>>() Line 712	C++
 	ucrtbase.dll!<lambda>(void)()	Unknown
 	ucrtbase.dll!__crt_seh_guarded_call<int>::operator()<<lambda_7777bce6b2f8c936911f934f8298dc43>,<lambda>(void) &,<lambda_3883c3dff614d5e0c5f61bb1ac94921c>>()	Unknown
 	ucrtbase.dll!_execute_onexit_table()	Unknown
 	torch_cpu.dll!dllmain_crt_process_detach(const bool is_terminating) Line 182	C++
 	torch_cpu.dll!dllmain_dispatch(HINSTANCE__ * const instance, const unsigned long reason, void * const reserved) Line 293	C++
 	ntdll.dll!LdrpCallInitRoutine()	Unknown
 	ntdll.dll!LdrpProcessDetachNode()	Unknown
 	ntdll.dll!LdrpUnloadNode()	Unknown
 	ntdll.dll!LdrpUnloadNode()	Unknown
 	ntdll.dll!LdrpDecrementModuleLoadCountEx()	Unknown
 	ntdll.dll!LdrUnloadDll()	Unknown
 	KernelBase.dll!FreeLibrary()	Unknown
 	[Inline Frame] Demo.exe!boost::dll::detail::shared_library_impl::unload() Line 120	C++
 	[Inline Frame] Demo.exe!boost::dll::detail::shared_library_impl::{dtor}() Line 39	C++
 	[Inline Frame] Demo.exe!boost::dll::shared_library::{dtor}() Line 179	C++
 	Demo.exe!test() Line 13	C++
 	Demo.exe!main() Line 17	C++
 	[External Code]	
```

which points to the destructor of static instance of `ProfileRegistry` class at torch/csrc/jit/runtime/profiling_record.cpp.

## Expected behavior

torch_cpu.dll and TorchClient using it both unload cleanly.

## Environment

 - PyTorch Version (e.g., 1.0): 1.8.1, 1.9.0
 - OS (e.g., Linux): Windows 10 21H1
 - How you installed PyTorch (`conda`, `pip`, source): source
 - Build command you used (if compiling from source):
```
cd g:\\builds\VSyWEYhe\0\asr3\libraries\libtorch/.cx2\15e39b\1 && cmake -G "Ninja" -DCONAN_LINK_RUNTIME="/MD" -DCMAKE_BUILD_TYPE="Release" -DCONAN_IN_LOCAL_CACHE="ON" -DCONAN_COMPILER="Visual Studio" -DCONAN_COMPILER_VERSION="16" -DCONAN_CXX_FLAGS="/MP8" -DCONAN_C_FLAGS="/MP8" -DCMAKE_INSTALL_PREFIX="g:\\builds\VSyWEYhe\0\asr3\libraries\libtorch/.cx2\934b91\1" -DCMAKE_INSTALL_BINDIR="bin" -DCMAKE_INSTALL_SBINDIR="bin" -DCMAKE_INSTALL_LIBEXECDIR="bin" -DCMAKE_INSTALL_LIBDIR="lib" -DCMAKE_INSTALL_INCLUDEDIR="include" -DCMAKE_INSTALL_OLDINCLUDEDIR="include" -DCMAKE_INSTALL_DATAROOTDIR="share" -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY="ON" -DCONAN_EXPORTED="1" -DCMAKE_PROJECT_INCLUDE="g:\\builds\VSyWEYhe\0\asr3\libraries\libtorch/.cx2\15e39b\1\conan_paths.cmake" -Dpybind11_PREFER_third_party="False" -DBUILD_CUSTOM_PROTOBUF="False" -DFBGEMM_LIBRARY_TYPE="static" -DINTEL_COMPILER_DIR="g:\builds\VSyWEYhe\0\asr3\libraries\libtorch\.cx2\fe0bb9\1\compiler" -DINTEL_MKL_DIR="g:\builds\VSyWEYhe\0\asr3\libraries\libtorch\.cx2\fe0bb9\1\mkl" -DBLAS="MKL" -DUSE_OPENMP="False" -DUSE_TBB="True" -DATEN_THREADING="TBB" -DMKL_THREADING="RT" -DMKLDNN_CPU_RUNTIME="TBB" -DBUILD_PYTHON="False" -DBUILD_TEST="False" -DUSE_CUDA="True" -DUSE_CUDNN="True" -DTORCH_CUDA_ARCH_LIST="Common" -DBUILD_SPLIT_CUDA="True" -DUSE_DISTRIBUTED="False" -DUSE_NCCL="False" -DUSE_NUMPY="False" -DUSE_KINETO="False" -DUSE_NNPACK="False" -DUSE_QNNPACK="False" -DUSE_PYTORCH_QNNPACK="False" -DUSE_XNNPACK="False" -DPSIMD_SOURCE_DIR="g:\builds\VSyWEYhe\0\asr3\libraries\libtorch\.cx2\15e39b\1\pytorch\third_party\psimd" -Wno-dev g:\\builds\VSyWEYhe\0\asr3\libraries\libtorch/.cx2\15e39b\1\pytorch
```
 - Python version: any
 - CUDA/cuDNN version: 11.2/8.1.1
 - GPU models and configuration: NVIDIA GeForce GTX 1650 Ti
 - Any other relevant information: MWE and libtorch built using VS 2019 16.10.2

## Additional context
* Actual `Access violation` when unloading torch_cpu.dll can be reproduced using regular account. It doesn't cause application crash though.
* Issue is reproducible with official build from pytorch.org https://download.pytorch.org/libtorch/cu111/libtorch-win-shared-with-deps-1.8.1%2Bcu111.zip

### PR DESCRIPTION
**Pull request by ghost**
It turns out on Windows static destructror should run on behalf the same library which run static constructor. For example, the objects registered by `torch_cuda_cpp.dll` should be properly deregistered by `torch_cuda_cpp.dll`, otherwise `torch_cpu.dll` will crash on unload.

Hopefully fixes #61111

### CODE REVIEW

File: torch/csrc/jit/runtime/profiling_record.cpp [hunk 1]
```diff
@@ -22,9 +22,16 @@ class ProfileRegistry {
     return &profile_registry_;
   }
 
-  void registerProfileNode(const std::function<bool(const Node*)>& func) {
+  size_t registerProfileNode(const std::function<bool(const Node*)>& func) {
     std::lock_guard<std::mutex> guard(mutex_);
-    registry_funcs_.push_back(func);
+    const auto key = registry_funcs_.size();
```
Comment (peterjc123) [inline torch/csrc/jit/runtime/profiling_record.cpp:27]: This is somehow dangerous to use the number of the dictionary as key. Suppose I call unregister and then register once, then the last function will be overridden.

**Review by peterjc123 — COMMENTED** [review]
Thanks, the changes look good from my perspective. But some CI tests on Linux didn't pass, maybe rebasing could help? BTW, maybe @ezyang @albanD can help to review this stuff.

**Comment by ezyang** [comment]
@jamesr66a do you know who the right person to review this is?

File: torch/custom_class_detail.h [hunk 2]
```diff
@@ -187,6 +187,21 @@ class TORCH_API class_base {
 
 TORCH_API void registerCustomClass(at::ClassTypePtr class_type);
 TORCH_API void registerCustomClassMethod(std::unique_ptr<jit::Function> method);
+TORCH_API void deregisterCustomClassMethod(const c10::QualifiedName& name);
+
+template<typename Unused = void>
+void deregisterCustomClassMethodOnExit(c10::QualifiedName name) {
```
[the remark this answers is not in the document]
Comment (dorozhkin-stc) [reply torch/custom_class_detail.h]: That was intended to take rvalue-only. In the newer version I have redesigned the deregistration facility and moved it to the `torch::class_` which is already a template.

File: torch/csrc/jit/runtime/profiling_record.cpp (hunk 1 above)
Comment (dorozhkin-stc) [reply torch/csrc/jit/runtime/profiling_record.cpp]: Thanks, I have come up with saner version which keeps increasing counter in a member field.

**Comment by dorozhkin-stc** [comment]
Rebased on recent master and cleaned up. Is there a windows build corresponding to my branch available online so I can double-check if the issue fixed?

**Comment by ezyang** [comment]
@dorozhkin-stc I went ahead and triggered binary builds for this PR, although I don't know if the binaries are saved in a spot you can get at (@zhouzhuojie do you know?)

**Comment by jamesr66a** [comment]
Does this have implications on custom class usage that should be mentioned in the documentation?

**Comment by dorozhkin-stc** [comment]
> Does this have implications on custom class usage that should be mentioned in the documentation?

None known. The intended static deregistration on library unload (which is the reverse of the static registration on library load) is designed to run completely automatically and transparent to the user.

**Comment by ezyang** [comment]
you can get the windows binaries on the "Artifacts" tab, see https://app.circleci.com/pipelines/github/pytorch/pytorch/354963/workflows/78d15278-8a89-47ad-96ee-2bbeeb3b8cb5/jobs/14962925/artifacts for example

**Review by jamesr66a — APPROVED** [review]
This seems fine to me. Adding @tugsbayasgalan @SplitInfinity @gmagogsfm as most proximate torchbind owners for additional review

**Comment by ezyang** [comment]
@jamesr66a will you land?

**Comment by jamesr66a** [comment]
@ezyang  sure

**Comment by jamesr66a** [comment]
@dorozhkin-stc can you rebase? Some of the test jobs seem to be failing due to infrastructure errors

**Comment by dorozhkin-stc** [comment]
After rebase there are still some job failures. Seems unrelated though (hard drive is out of free space on Windows machine and deadline timeout on the others).

**Review by gmagogsfm — APPROVED** [review]

**Comment by gmagogsfm** [comment]
Looks good, the CI test failures seem unrelated to your change. Could you rebase to trigger a re-run so that real issues (if any) don't get masked by those unrelated failures? Thanks!

**Comment by DBraun** [comment]
Is there an update on this? I would really like to use a newer version of libtorch, and I think this issue is preventing me.

**Comment by jamesr66a** [comment]
@dorozhkin-stc it looks like there's still infra failures on CI here, can you rebase again?

**Comment by ezyang** [comment]
bumping

**Pull request closed without merging**